### PR TITLE
fix building overlap for isometric view (mostly)

### PR DESCRIPTION
**Pull request by michaelkirk**
**Before**

In the current isometric rendering, there are perspective issues with buildings being rendered on top of eachother in undesirable ways. And note the car parked on the building roof.

Screen Shot 2020-09-18 at 4 19 54 PM

Disregard that some of the building's groundfloors juts into the street. As you can see here, it's present in the default rendering, before these changes, and seems to just be bad OSM data:

Screen Shot 2020-09-18 at 4 19 33 PM

**After**

Most buildings look good after these changes.

Screen Shot 2020-09-18 at 4 17 13 PM

You can see the z-indexes make sense for agents traveling behind the buildings too, as far as perspective goes. But as discussed previously, this might undesirably obscure details of the simulation. Future work which might address this could be some level of transparency, or rotation, or something else.

iso_buildings mov

## math

The thrust of this change is to offset the z-value**, based on it's distance from the orthographic axis. I've attached a drawing which may or may not help clarify:

ortho_math

**futzing with z-indices is fraught with peril, but I'm not sure how else to address something like this.  My proposal for enforcing a modicum of sanity is:

We already have some course grained base app layers which have an associated z-index.
1. map
2. menu
3. tooltips
4. etc.

The z-value for these layers should be whole integer values.

Any adjustment to z, like we do for building height, is applied to the base layer, and must not exceed 1.0, so that way we don't inadvertently traverse layers, e.g. we don't want a really tall building to be rendered "above" a tooltip.

Like I said, it's kind of fraught, but it's the best I could come up with. 🤷

I'm not married to this approach if you have other ideas, or just want to hold off on merging for now.

### CODE REVIEW

File: widgetry/src/geom.rs [hunk 1]
```diff
@@ -7,7 +7,7 @@ use geom::{Angle, Bounds, Polygon, Pt2D};
 /// A mutable builder for a group of colored polygons.
 #[derive(Clone)]
 pub struct GeomBatch {
-    pub(crate) list: Vec<(Fill, Polygon)>,
+    pub(crate) list: Vec<(Fill, Polygon, f64)>,
```
Comment (dabreegster) [inline widgetry/src/geom.rs:10]: nit: Add a comment like `// f64 is the Z value`?

File: widgetry/src/shaders/vertex_140.glsl [hunk 2]
```diff
@@ -25,9 +25,10 @@ void main() {
     // Translate that to normalized device coordinates (NDC)
     float x = (screen_x / window[0] * 2.0) - 1.0;
     float y = (screen_y / window[1] * 2.0) - 1.0;
+    float z = position[2] + window[2];
```
Comment (dabreegster) [inline widgetry/src/shaders/vertex_140.glsl:28]: Ah, this is the part that was confusing me. Makes perfect sense -- lots of times, we want to conveniently set a base Z value in the uniforms, but now we're adding the ability to tune this per polygon. Could we add a comment to that effect in `drawing.rs` by the Z value definitions, and mention keeping the polygon Z value under 1.0 (for the reasons you had in the PR description)?

**Review by dabreegster — COMMENTED** [review]
This looks great! I tested in a bunch of maps and can't really spot too many weird things. I'm also excited about the per-polygon Z value overrides; I can't remember now, but every now and then I've hit a case where they would've been handy.

> Disregard that some of the building's groundfloors just into the street.

This mostly comes from bad inference of parking lanes on roads that haven't been explicitly tagged yet. I've thought about autodetecting these cases and suggesting parking edits to upstream in OSM.

**Comment by michaelkirk** [comment]
Ok, I've added some docs and a debug_assert, PTAL @dabreegster

**Comment by dabreegster** [comment]
LGTM, thanks for turning this quick experiment into something solid!

**Pull request merged**